polly/src/pollable.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+use std::fmt::Formatter;
+use std::os::unix::io::{AsRawFd, RawFd};
+use std::rc::{Rc};
+use std::ops::Deref;
+
+#[derive(Default, Clone, PartialEq)]
+pub struct OwnedFD {
```

Comment:
This doesn't help with guaranteeing `fd` validity/lifetime. This wrapper doesn't impose any lifetime restrictions on the real `fd` owner which can go out of scope/close the `fd` while we're still happily using it here.

I suggest polly takes something like `Rc<dyn AsRawFd>` and is given a counted reference to the real object backing the `fd`. That is the only way I can think of to guarantee fd correctness.

The alternative is to just go back to `RawFd` and let the user manage their logic as to keep things safe.

Reply:
I've implemented the dup solution since the Rc<dyn AsRawFd> option did not work due to double mutable borrowing. 

EventHandlers should take care of closing of the original fd.

Reply:
> I think there is a solution with `OwnedFd` that doesn't imply wrapping `Rc`:
> 
> * `libc::dup2` the original `fd` and wrap the duplicate in this `OwnedFd`. We'd have to confirm that `EPOLL*` events are mirrored on duplicated `fd`s.
> 
> This moves the reference counted lifetime of the `file` to be handled by the kernel instead of our app, **but** might lead to unexpected behavior as seen by a user of the crate. For example:
> 
> > Open a socket, bind it to some addr/port, use it in polly then close it (without removing from polly).
> 
> Expected behavior is for the `addr/port` combo to now be available, but they won't because the socket hasn't closed yet, polly is keeping it alive.

Not seems a good idea by dup2, we also need to take care of duplicated fds and their flags (CLO_EXEX etc) when live upgrading.

Reply:
> This doesn't help with guaranteeing `fd` validity/lifetime. This wrapper doesn't impose any lifetime restrictions on the real `fd` owner which can go out of scope/close the `fd` while we're still happily using it here.
> 
> I suggest polly takes something like `Rc<dyn AsRawFd>` and is given a counted reference to the real object backing the `fd`. That is the only way I can think of to guarantee fd correctness.
> 
> The alternative is to just go back to `RawFd` and let the user manage their logic as to keep things safe.

Prefer the way for the user to manage RawFd's lifetime.

Reply:
@jiangliu thanks for the feedback.
I have updated this PR and removed the duplication of fds while I was experimenting with refactoring the block device to implement EventHandler trait. 

Please take a look at the updated commit.